### DISCUSSION TITLE
Map Enum to Null when not matched by Name

### DISCUSSION THREAD

**Discussion by matthiaslischka** [AutoMapper/AutoMapper.Extensions.EnumMapping#14] [Q&A]
I have a nullable enum property on the source side and get this property with value 0 instead of NULL.
But 0 is not a represented value in the enum so AutoMapper with `ConvertUsingEnumMapping` throws an exception that the enum can not be mapped. Is there a way to map it to `NULL` whenever the value is not represented in the source enum? Some way to prevent this runtime error?

```csharp
public class SomeSourceDto
{
    public SomeSourceEnum? SomeEnum { get; set; }
}

public class SomeDestinationDto
{
    public SomeDestinationEnum? SomeEnum { get; set; }
}

public enum SomeSourceEnum
{
    First = 1
}

public enum SomeDestinationEnum
{
    First = 1
}
```

Mapping configuration:
```csharp
CreateMap<SomeSourceDto, SomeDestinationDto>();
CreateMap<SomeSourceEnum, SomeDestinationEnum>().ConvertUsingEnumMapping(opt => opt.MapByName());
```

Mapping:
```csharp
var someSourceDto = new SomeSourceDto { SomeEnum = 0 };
Mapper.Map<SomeDestinationDto>(someSourceDto);
```

Exception:

> AutoMapper.AutoMapperMappingException
> 
> Error mapping types.
> 
> ...
> 
> Destination Member:
> 
> SomeEnum

My current workaround is to apply this before the map:
```csharp
someSourceDto.SomeEnum = someSourceDto.SomeEnum == 0 ? null : someSourceDto.SomeEnum;
```

Is there a better way? I can't use `ConvertUsingEnumMapping` when I make the enum types in the map nullable.

The problem is that this can happen for every enum property so I'm at big risk of runtime errors atm because I have no control over the source side of the data.

Example repo: https://github.com/matthiaslischka/AutoMapperNullEnumIssue/blob/main/AutoMapperNullEnumIssue/UnitTest1.cs

**Comment by lbargaoanu** [comment]
You can write that check inside a `MapFrom`.

**Reply by matthiaslischka** [reply]
Isn't there a cleaner way when the property is already nullable? Some fallback to null when the ConvertUsingEnumMapping does not yield a result?

I would have to do this for every nullable enum value to prevent unnecessary runtime errors. And probably not only for value but for all unmapped values.

But then I also have an enum where I can map 9 out of 10 by name and the 10th needs a manual map because there is a typo:
```chsarp
opt.MapByName().MapValue(SomeSourceEnum.FoBar, SomeDestinationEnum.FooBar);
```

Would be nice if ConvertUsingEnumMapping could be applied to nullable Enums and map to null if no map was found but that does not work atm.

**Reply by lbargaoanu** [reply]
Your case seems totally arbitrary, but I have no thoughts on `ConvertUsingEnumMapping`.

**Reply by matthiaslischka** [reply]
How is it arbitrary? ^^ Mapping a nullable enum property to null when the map does not yield an result instead of throwing an runtime error doesn't sound so far fetched, does it?

**Reply by lbargaoanu** [reply]
It certainly doesn't sound like code that belongs in a reusable library, but for app code, sure.

**Reply by matthiaslischka** [reply]
Dunno. I'm actually wondering if there isn't _already_ a way to handle some default/fallback behavior in the library that I'm just overlooking.

We use `ConvertUsingEnumMapping` and `MapByName()` in order to find any [missing enum mappings](https://github.com/matthiaslischka/AutoMapperNullEnumIssue/blob/700d1b3ecd5c744aaa1fe6f80210fd972c274985/AutoMapperNullEnumIssue/UnitTest1.cs#L37) at unittest time with `EnableEnumMappingValidation` and `AssertConfigurationIsValid`. But this runtime error when the enum value is undefined - surely I'm not the first one running into this.

**Comment by HenkKin** [comment]
See issue: https://github.com/AutoMapper/AutoMapper.Extensions.EnumMapping/issues/4#issuecomment-673387878
